### DISCUSSION TITLE
Specify a custom Firefox path with a new command in script

### DISCUSSION THREAD

**Discussion by ruihildt** [wxt-dev/wxt#658] [Q&A]
Hi,

I'm working with a project where I need to test my extension on both Firefox and Mullvad Browser (a Firefox fork).

I successfully launch it by doing the following:

``` JavaScript
runner: {
    binaries: { firefox: '/home/rui/apps/mullvad-browser/Browser/mullvadbrowser' },
  },
```

But it means I can't test Firefox without manually changing this config every time I want to test another channel of a Firefox based browser.

I tried to add a custom entry like this:

``` JavaScript
runner: {
    binaries: { mullvad: '/home/rui/apps/mullvad-browser/Browser/mullvadbrowser' },
  },
```

Invoking from the script with `wxt -b mullvad --mv2` correctly launch the browser, but the extension is not loaded and some tabs open with extension related URIs (which is not surprising since it doesn't know which browser it is).

What would be the best way to have multiple versions of Firefox being one script away from testing?

**Comment by aklinker1** [comment]
Yeah, this is something I've put off for a while lol. I need to add some way for you to map from a `--browser` name (which can be any string) to a big 3 browser (chrome, firefox, safari).

Thoughts on a config like this? 

```ts
export default defineConfig({
  targets: {
    mullvad: "firefox",
    ios: "safari",
  },
})
```

The default would probably be this:

```ts
{
  chrome: "chrome",
  chromium: "chrome",
  edge: "chrome",
  opera: "chrome",
  firefox: "firefox",
  mullvad: "firefox",
  safari: "safari",
  ios: "safari",
}
```

**Reply by ruihildt** [reply]
I guess that would mean I could do something like this:

``` JavaScript
export default defineConfig({
  targets: {
  mullvad: "firefox",
  firefoxNightly:  "firefox",
  },
  runner: {
    binaries: { 
        mullvad: '/home/rui/apps/mullvad-browser/Browser/mullvadbrowser',
        firefoxNightly: '/home/rui/apps/firefox-nightly/Browser/firefox-bin', 
    },
  },
})
```

If that's the case, then it's a bit weird to have to declare a custom target and use that target name in the binaries section in the same config.

Maybe, could we merge all custom properties in a browser config like this:

``` JavaScript
export default defineConfig({
  runner: {
        mullvad: {
          binary: '/home/rui/apps/mullvad-browser/Browser/mullvadbrowser',
          target: 'firefox',
          startUrls: [] // and other such properties
         },
        nightly: {
          binary: '/home/rui/apps/firefox-nightly/Browser/firefox-bin',
          target: 'firefoxnightly'
       },
       firefox: { // to override or customize the default "firefox" command behavior
          startUrls: []
       }
  },
})
```

But I've never architectured such projects, so maybe this is not really a good way to do things.^^"
I would be perfectly fine to use your suggestion as well. :)

**Reply by ruihildt** [reply]
@aklinker1 Pinging you here to remind you about this and check if you've had further thoughts regarding my proposal.

**Reply by aklinker1** [reply]
Haven't forgotten about it! It's in my work queue. I was gonna look more into your proposal when I start working on this feature

https://github.com/orgs/wxt-dev/projects/2/views/1
